dff/core/pipeline/pipeline/pipeline.py

```diff
@@ -0,0 +1,211 @@
+            or before/after each pipeline component.
+        They can be used for pipeline statistics collection or other functionality extensions.
+        NB! Global wrappers are still wrappers,
+            they shouldn't be used for much time-consuming tasks (see ../service/wrapper.py).
+        :global_wrapper_type: (required) - GlobalWrapperType indication where the wrapper function should be executed.
+        :wrapper: (required) - wrapper function itself.
+        :whitelist: - a list of services to only add this wrapper to.
+        :blacklist: - a list of services to not add this wrapper to.
+        Returns None.
+        """
+
+        def condition(name: str) -> bool:
+            return (whitelist is None or name in whitelist) and (blacklist is None or name not in blacklist)
+
+        if (
+            global_handler_type is GlobalExtraHandlerType.BEFORE_ALL
+            or global_handler_type is GlobalExtraHandlerType.AFTER_ALL
+        ):
+            whitelist = ["pipeline"]
+            global_handler_type = (
+                GlobalExtraHandlerType.BEFORE
+                if global_handler_type is GlobalExtraHandlerType.BEFORE_ALL
+                else GlobalExtraHandlerType.AFTER
+            )
+
+        self._services_pipeline.add_extra_handler(global_handler_type, extra_handler, condition)
+
+    @property
+    def info_dict(self) -> dict:
+        """
+        Property for retrieving info dictionary about this pipeline.
+        Returns info dict, containing most important component public fields as well as its type.
+        All complex or unserializable fields here are replaced with 'Instance of [type]'.
+        """
+        return {
+            "type": type(self).__name__,
+            "messenger_interface": f"Instance of {type(self.messenger_interface).__name__}",
+            "context_storage": f"Instance of {type(self.context_storage).__name__}",
+            "services": [self._services_pipeline.info_dict],
+        }
+
+    def pretty_format(self, show_extra_handlers: bool = False, indent: int = 4) -> str:
+        """
+        Method for receiving pretty-formatted string description of the pipeline.
+        Resulting string structure is somewhat similar to YAML string.
+        Should be used in debugging/logging purposes and should not be parsed.
+        :show_wrappers: - whether to include Wrappers or not (could be many and/or generated).
+        :indent: - offset from new line to add before component children.
+        """
+        return pretty_format_component_info_dict(self.info_dict, show_extra_handlers, indent=indent)
+
+    @classmethod
+    def from_script(
+        cls,
+        script: Union[Script, Dict],
+        start_label: NodeLabel2Type,
+        fallback_label: Optional[NodeLabel2Type] = None,
+        context_storage: Optional[Union[DBAbstractConnector, Dict]] = None,
+        messenger_interface: MessengerInterface = CLIMessengerInterface(),
```

Comment:
Why should we do it? maybe you can see some pros?

Reply:
The difference is that now the `CLIMessengerInterface` instance is created once the class is first visited and this instance is used as a default argument for all function calls.
That might become inconvenient if user creates Pipeline, alters any fields of it's messenger interface and then after creating a new Pipeline it will have not a brand new messenger interface but the altered one.